### PR TITLE
0-1 Knapsack

### LINKED ISSUE #188: The 0-1 Knapsack Problem.
I would like to add the 0-1 Knapsack problem, a very famous Dynamic Programming algorithm.

### PR DESCRIPTION
**Pull request by heisenberg-2505**
Implemented the 0-1 knapsack problem.

- [x] Read the [contribution guidelines][contrib-guidelines]
- [x] Added [unit tests][unit-tests] (or unit tests have already been added)
- [x] Updated the [Contents][contents]

Closes #188 

[contrib-guidelines]: https://github.com/faheel/Algos/blob/master/CONTRIBUTING.md
[unit-tests]: https://github.com/faheel/Algos/blob/master/C%2B%2B/UNIT_TESTS.md
[contents]: https://github.com/faheel/Algos/tree/master/C%2B%2B#contents

### CODE REVIEW

File: C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp]: [Avoid `using namespace std`](https://stackoverflow.com/questions/1452721/why-is-using-namespace-std-considered-bad-practice). Instead, reference what you need specifically, as in: 

`using std::vector;`

File: C++/test/algorithm/dynamic_programming/0_1_knapsack.cpp
Comment (alxmjo) [inline C++/test/algorithm/dynamic_programming/0_1_knapsack.cpp]: Avoid `using namespace std`.

File: C++/test/algorithm/dynamic_programming/0_1_knapsack.cpp
Comment (alxmjo) [inline C++/test/algorithm/dynamic_programming/0_1_knapsack.cpp]: Can you add a few more test cases?

**Review by alxmjo — CHANGES_REQUESTED** [review]
Looks good! Just requesting a few little changes here.

File: C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp]: If code isn't being used, remove it rather than just commenting it out.

File: C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp]: Instead of ll for long long, use LL. ll is very difficult to distinguish from 11.

**Review by alxmjo — CHANGES_REQUESTED** [review]
I'd also like @Faheel to have a look at this.

**Review by alxmjo — APPROVED** [review]

**Comment by alxmjo** [comment]
Thanks for making those changes. Going to leave this up to @faheel to review and merge.

**Comment by alxmjo** [comment]
@heisenberg-2505 While we're waiting for Faheel's review, would you like to update the [Contents](https://github.com/faheel/Algos/blob/reorganise/C%2B%2B/README.md)?

File: C++/test/algorithm/dynamic_programming/0_1_knapsack.cpp
Comment (faheel) [inline C++/test/algorithm/dynamic_programming/0_1_knapsack.cpp]: You can add
```c++
using std::vector;
```
to prevent writing prefixing `vector` with `std::` every time.

File: C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp
Comment (faheel) [inline C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp]: * Having whitespace between `std` `::` `vector` `<` `LL` `>` etc. isn't recommended. The following is better:
  ```c++
  std::vector<std::vector<LL>> K(numberOfItems + 1, std::vector<LL>(capacity + 1));
  ```
  Same for the function's parameters.
* `K` isn't a very informative name. Something like `maximum` or `maxValue` would be better.

File: C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp
Comment (faheel) [inline C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp]: `KNAPSACK_HPP` isn't very specific and might conflict with other knapsack implementations that would be added later, so please change it to `0_1_KNAPSACK_HPP`.

**Review by faheel — CHANGES_REQUESTED** [review]
Please also add an entry for 0-1 knapsack in the contents (in `C++/README.md).

File: C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp
Comment (heisenberg-2505) [reply C++/include/algorithm/dynamic_programming/0_1_knapsack.hpp]: When using `0_1_KNAPSACK_HPP` it gives the following error `macro names must be identifiers` so using `KNAPSACK_0_1_HPP` instead.

**Comment by heisenberg-2505** [comment]
Can you review my pull request?

**Review by faheel — APPROVED** [review]

**Comment by faheel** [comment]
@heisenberg-2505 Thank you for contributing! :+1:

**Pull request merged**
